# gridss

performs somatic genomic rearrangement detection and classification

## Overview

 GRIDSS is a module software suite containing tools useful for the detection of 
 genomic rearrangements. GRIDSS includes a genome-wide break-end assembler, as 
 well as a structural variation caller for Illumina sequencing data. GRIDSS 
 calls variants based on alignment-guided positional de Bruijn graph genome-wide
 break-end assembly, split read, and read pair evidence.
 
 GRIDSS makes extensive use of the standard tags defined by SAM specifications. 
 Due to the modular design, any step (such as split read identification) can be
 replaced by another implementation that also outputs using the standard tags. 
 It is hoped that GRIDSS can serve as an exemplar modular structural variant 
 pipeline designed for interoperability with other tools.

## Dependencies

* [GRIDSS](https://github.com/PapenfussLab/gridss)
* [hmftools](https://github.com/hartwigmedical/hmftools)


## Usage

### Cromwell
```
java -jar cromwell.jar run gridss.wdl --inputs inputs.json
```

### Inputs

#### Required workflow parameters:
Parameter|Value|Description
---|---|---
`tumorBam`|File|Input tumor file (bam)
`tumorBai`|File|Input tumor file index (bai)
`normBam`|File|Input normal file (bam)
`normBai`|File|Input normal file index (bai)


#### Optional workflow parameters:
Parameter|Value|Default|Description
---|---|---|---
`assemblyChunks`|Int|4|How many chunks to use for assembly job, may be calculated based on input size
`outputFileNamePrefix`|String|basename("~{tumorBam}",".filter.deduped.realigned.recalibrated.bam")|Output file prefix


#### Optional task parameters:
Parameter|Value|Default|Description
---|---|---|---
`preprocessNormal.blacklist`|String?|None|bed file with regions to ignore
`preprocessNormal.modules`|String|"gridss/2.13.2 hmftools-data/hg38"|Required environment modules
`preprocessNormal.refFasta`|String|"$HMFTOOLS_DATA_ROOT/hg38_random.fa"|Reference FASTA file
`preprocessNormal.gridssScript`|String|"$GRIDSS_ROOT/gridss --jar $GRIDSS_ROOT/gridss-2.13.2-gridss-jar-with-dependencies.jar"|Script to run GRIDSS
`preprocessNormal.workingDir`|String|"~{basename(inputBam)}.gridss.working"|Working dir for processing, automatically generated by gridds
`preprocessNormal.memory`|Int|16|Memory allocated for this job (GB)
`preprocessNormal.timeout`|Int|12|Hours before task timeout
`preprocessNormal.threads`|Int|4|Requested CPU threads
`preprocessTumor.blacklist`|String?|None|bed file with regions to ignore
`preprocessTumor.modules`|String|"gridss/2.13.2 hmftools-data/hg38"|Required environment modules
`preprocessTumor.refFasta`|String|"$HMFTOOLS_DATA_ROOT/hg38_random.fa"|Reference FASTA file
`preprocessTumor.gridssScript`|String|"$GRIDSS_ROOT/gridss --jar $GRIDSS_ROOT/gridss-2.13.2-gridss-jar-with-dependencies.jar"|Script to run GRIDSS
`preprocessTumor.workingDir`|String|"~{basename(inputBam)}.gridss.working"|Working dir for processing, automatically generated by gridds
`preprocessTumor.memory`|Int|16|Memory allocated for this job (GB)
`preprocessTumor.timeout`|Int|12|Hours before task timeout
`preprocessTumor.threads`|Int|4|Requested CPU threads
`assembleBam.blacklist`|String?|None|bed file with regions to ignore
`assembleBam.modules`|String|"gridss/2.13.2 hmftools-data/hg38"|Required environment modules
`assembleBam.refFasta`|String|"$HMFTOOLS_DATA_ROOT/hg38_random.fa"|Reference FASTA file
`assembleBam.gridssScript`|String|"$GRIDSS_ROOT/gridss --jar $GRIDSS_ROOT/gridss-2.13.2-gridss-jar-with-dependencies.jar"|Script to run GRIDSS
`assembleBam.workingDirNorm`|String|"~{basename(normBam)}.gridss.working"|Working dir for normal bam, automatically generated by gridds
`assembleBam.workingDirTumr`|String|"~{basename(tumorBam)}.gridss.working"|Working dir for tumor bam, automatically generated by gridds
`assembleBam.memory`|Int|32|Memory allocated for this job (GB)
`assembleBam.timeout`|Int|24|Hours before task timeout
`assembleBam.threads`|Int|8|Requested CPU threads
`callSvs.workingDirNorm`|String|"~{basename(normBam)}.gridss.working"|Working dir for normal bam, automatically generated by gridds
`callSvs.workingDirTumr`|String|"~{basename(tumorBam)}.gridss.working"|Working dir for tumor bam, automatically generated by gridds
`callSvs.blacklist`|String?|None|bed file with regions to ignore
`callSvs.modules`|String|"gridss/2.13.2 hmftools/1.0 hmftools-data/hg38"|Required environment modules
`callSvs.refFasta`|String|"$HMFTOOLS_DATA_ROOT/hg38_random.fa"|Reference FASTA file
`callSvs.gridssScript`|String|"$GRIDSS_ROOT/gridss --jar $GRIDSS_ROOT/gridss-2.13.2-gridss-jar-with-dependencies.jar"|Script to run GRIDSS
`callSvs.threads`|Int|8|Requested CPU threads
`callSvs.memory`|Int|50|Memory allocated for this job (GB), split b/w high-demand chunk and everything else
`callSvs.overhead`|Int|8|Memory for other things ran by JVM
`callSvs.timeout`|Int|24|Hours before task timeout


### Outputs

Output | Type | Description
---|---|---
`structuralVcf`|File|Structural Variant .vcf file


## Commands
 This section lists commands run by gridss workflow
 
 * Running GRIDSS
 
 perform joint SV calling on the tumour and normal. The workflow consists of three steps
 described below:
 
### Preprocess
 
 This is a step of preprocessing inputs
 
### Call SVs
 
 Using pre-processed data call Structural Variants
 
```
    GRIDSS_SCRIPT 
    -b BLACKLIST \
    -r REF_FASTA \
    -s preprocess \
    -t THREADS \
    INPUT_BAM
```
 
### Assembly
 
 Assembly is performed in parallel on 10Mbase chunks. This step scales to 8 cores with peak memory usage of 31Gb
 and can be distributed across multiple nodes (using scatter)
 
```
     set -euo pipefail
     mkdir WORKDIR_NORM WORKDIR_TUMOR
     ln -s PROCESSED_NORMBAM -t WORKDIR_NORM
     ln -s PROCESSED_NORMIDX -t WORKDIR_NORM
     ln -s PROCESSED_TUMRBAM -t WORKDIR_TUMOR
     ln -s PROCESSED_TUMRIDX -t WORKDIR_TUMOR
 
     GRIDSS_SCRIPT
     -b BLACKLIST \
     -r REF_FASTA \
     -t THREADS \
     -s assemble \
     -a assembly.bam \
     --jobnodes JOB_NODES \
     --jobindex JOB_INDEX \
     NORM_BAM TUMOR_BAM
```
 
### Call SVs
 
 Using pre-processed data call Structural Variants
 
```
     set -euo pipefail
     mkdir WORKDIR_NORM WORKDIR_TUMOR
     ln -s PROCESSED_NORMBAM -t WORKDIR_NORM
     ln -s PROCESSED_NORMIDX -t WORKDIR_NORM
     ln -s PROCESSED_TUMRBAM -t WORKDIR_TUMOR
     ln -s PROCESSED_TUMRIDX -t WORKDIR_TUMOR
     mkdir assembly.bam.gridss.working
     for f in ~{sep=' ' assembleChunks}
     do
       ln -s $f -t assembly.bam.gridss.working
     done
 
     GRIDSS_SCRIPT
     -b BLACKLIST \
     --jvmheap (MEMORY - OVERHEAD)g \
     --otherjvmheap OVERHEADg \
     --reference REF_FASTA \
     -a assembly.bam \
     -s assemble,call \
     -t THREADS \
     -o OUTPUT_PREFIX.allocated.vcf \
     NORM_BAM TUMOR_BAM
```

## Support

For support, please file an issue on the [Github project](https://github.com/oicr-gsi) or send an email to gsi@oicr.on.ca .

_Generated with generate-markdown-readme (https://github.com/oicr-gsi/gsi-wdl-tools/)_
